src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriterCache.cs

```diff
@@ -0,0 +1,96 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+
+namespace System.Text.Json
+{
+    /// <summary>
+    /// Defines a thread-local cache for JsonSerializer to store reusable Utf8JsonWriter/IBufferWriter instances.
+    /// </summary>
+    internal static class Utf8JsonWriterCache
+    {
+        [ThreadStatic]
+        private static ThreadLocalState? s_threadLocalState;
+
+        public static Utf8JsonWriter RentWriterAndBuffer(JsonSerializerOptions options, out PooledByteBufferWriter bufferWriter)
+        {
+            ThreadLocalState state = s_threadLocalState ??= new();
+            Utf8JsonWriter writer;
+
+            if (state.RentedWriters++ == 0)
+            {
+                // First JsonSerializer call in the stack -- initialize & return the cached instances.
+                bufferWriter = state.BufferWriter;
+                writer = state.Writer;
+
+                bufferWriter.InitializeEmptyInstance(options.DefaultBufferSize);
+                writer.Reset(bufferWriter, options.GetWriterOptions());
+            }
+            else
+            {
+                // We're in a recursive JsonSerializer call -- return fresh instances.
+                bufferWriter = new PooledByteBufferWriter(options.DefaultBufferSize);
```

Comment:
Where is the PooledByteBufferWriter getting its buffers from?  Does this mean we're potentially taking buffers that were available via ArrayPool to any thread and instead making them available only to this particular thread?